core/src/main/java/org/subtlelib/poi/impl/row/RowContextImpl.java

```diff
@@ -294,4 +303,41 @@ public RowContext totals(Formula formula, int times, Style style) {
         }
         return this;
     }
-}
\ No newline at end of file
+
+    @Override
+    public RowContext setFilterDataRange(FilterDataRange data) {
+        if (!data.isEndColMarked()) {
+            data.endOnPreviousCol();
+        }
+        if (!data.isEndRowMarked()) {
+            data.endOnPreviousRow();
+        }
+        this.filterData = data;
+        return this;
+    }
+
+    @Override
+    public RowContext filter() {
+        checkState(filterData != null, "Please set filter data range before applying filter (setFilterDataRange(...)");
+
+        CellRangeAddress filterRange = new CellRangeAddress(
+                filterData.getStartRowNo(),
+                filterData.getEndRowNo(),
+                filterData.getStartColNo(),
+                filterData.getEndColNo()
+        );
+        sheet.getNativeSheet().setAutoFilter(filterRange);
+
+        // auto-set column widths
+        // NOTE: see notice in http://poi.apache.org/spreadsheet/quick-guide.html#Autofit
+        // NOTE: To calculate column width Sheet.autoSizeColumn uses Java2D classes that throw exception if graphical environment is not available. In case if graphical environment is not available, you must tell Java that you are running in headless mode and set the following system property: java.awt.headless=true . You should also ensure that the fonts you use in your workbook are available to Java.
+        GraphicsEnvironment ge = GraphicsEnvironment.getLocalGraphicsEnvironment();
+        if (!ge.isHeadless()) {
+            for (int i = filterData.getStartColNo(); i <= filterData.getEndColNo(); i++) {
+                sheet.getNativeSheet().autoSizeColumn(i);
```

Comment:
autosizing columns has nothing to do with filtering and should be a separate thing.